### ISSUE TITLE
Detect invalid variable/expression usage with missing curly braces

### ISSUE THREAD

**Issue by abitrolly** [terraform-linters/tflint#436]
Currently, `tflint` is unable to find cases when people forget to use `{` `}` to render variables.
```
  triggers = {
    template_rendered = "$data.template_file.inventory_template.rendered"
  }
```
This code is wrong, because `"$data.template_file.inventory_template.rendered"` is treated as literal string, and it should be written as "`${data.template_file.inventory_template.rendered}`".

* [Invalid Terraform report](https://github.com/hashicorp/terraform/issues/21211#issuecomment-510703355)
* [Original bad code](https://github.com/abitrolly/XE3Bot/commit/ee4cd4b7e3fa9a61ac0f3c030146fee39f389929#diff-0e98d84c23e702e3baa05095730661b0)

**Comment by wata727** [comment]
Umm, this seems a little difficult.
First of all, it's not clear what the invalid variable/expression is. For instance, the above example is clearly a valid string literal. Is the following expression invalid? Or is it a string literal?

```hcl
value = "This is var.number colors"
```

```hcl
value = "This is $var.number colors"
```

Although we may be able to create regular expressions that can detect them well, we may need to be very careful with the implementation.

**Comment by abitrolly** [comment]
Does `tflint` maintain the list of variables found in `.tf` script? It could match $ regexp against that list.

**Comment by wata727** [comment]
Yes, TFLint has a list of variables and resources as a result of parsing.

https://github.com/wata727/tflint/blob/1c31b742deb0a556520dec70ae9d4da5740f2b48/tflint/runner.go#L26

https://github.com/hashicorp/terraform/blob/e97ad4ebef504883229a1c299be617dfb7bec017/configs/module.go#L32-L39

**Comment by abitrolly** [comment]
I understand that there should be a file in https://github.com/wata727/tflint/tree/master/rules/terraformrules that traverses all expanded string literals, then gets all substrings matching the regexp `\$\w[.\w]` and then check those list against the list of variables and give a warning. Is that right?

**Comment by wata727** [comment]
That's right. I think that it is the most difficult to implement a parser to search for "reference-like" from a string literal.

**Comment by abitrolly** [comment]
The parser is just a simple `regex` matching, no? https://play.golang.org/p/Nu4ZapS272N

**Comment by wata727** [comment]
Umm, probably it will work. However, I don't fully understand the definition of HCL expressions, so I can't ensure that this regular expression is perfectly correct.
To avoid confusion, it might be a good idea to add the rule without enabling by default if you want to add it.

**Comment by abitrolly** [comment]
This doesn't touch HCL expressions, because it works with plain text and matches it against HCL variables only.

**Comment by wata727** [comment]
PR was already closed https://github.com/terraform-linters/tflint/pull/469. Closing this.
